### PR TITLE
Darwinia <> Darwinia Parachain

### LINKED ISSUE #951: Darwinia <> Darwinia Parachain

### PR DESCRIPTION
**Pull request by AurevoirXavier**
Close #951.

Signed-off-by: Xavier Lau <xavier@inv.cafe>

### CODE REVIEW

File: runtime/darwinia/src/pallets/to_parachain_backing_.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+frame_support::parameter_types! {
+	pub const DarwiniaParachainChainId: ChainId = DARWINIA_PARACHAIN_CHAIN_ID;
+	pub const S2sBackingPalletId: PalletId = PalletId(*b"da/s2sba");
+	pub const MaxLockRingAmountPerTx: Balance = 1_000_000 * COIN;
+	pub const BridgeDarwiniaParachainLaneId: LaneId = DARWINIA_DARWINIA_PARACHAIN_LANE;
+}
+
+impl Config for Runtime {
```
Comment (HackFisher) [inline runtime/darwinia/src/pallets/to_parachain_backing_.rs:33]: Using new approach instead cc @xiaoch05

File: runtime/darwinia/src/pallets/bridge_dispatch.rs [hunk 2]
```diff
@@ -126,3 +128,15 @@ impl Config<WithCrabDispatch> for Runtime {
...
+impl Config<WithDarwiniaParachainDispatch> for Runtime {
+	type AccountIdConverter = bp_darwinia_parachain::AccountIdConverter;
+	type BridgeMessageId = (LaneId, MessageNonce);
+	type Call = Call;
+	type CallValidator = CallValidator;
+	type EncodedCall = bm_darwinia_parachain::FromDarwiniaParachainEncodedCall;
+	type Event = Event;
+	type IntoDispatchOrigin = IntoDispatchOrigin;
+	type SourceChainAccountId = bp_darwinia_parachain::AccountId;
+	type TargetChainAccountPublic = bp_crab::AccountPublic;
```
Comment (AsceticBear) [inline runtime/darwinia/src/pallets/bridge_dispatch.rs:140]:
```rs
type TargetChainAccountPublic = bp_crab::AccountPublic;
type TargetChainSignature = bp_crab::Signature;
```
`bp_crab` should be `bp_darwinia_parachain` ?

File: runtime/darwinia/Cargo.toml [hunk 3]
```diff
@@ -19,23 +19,27 @@ serde_json        = { version = "1.0", optional = true }
...
 # darwinia-network frame
 darwinia-balances        = { default-features = false, git = "https://github.com/darwinia-network/darwinia-common", branch = "darwinia-v0.12.5" }
 darwinia-ecdsa-authority = { default-features = false, git = "https://github.com/darwinia-network/darwinia-common", branch = "darwinia-v0.12.5" }
 darwinia-header-mmr      = { default-features = false, git = "https://github.com/darwinia-network/darwinia-common", branch = "darwinia-v0.12.5" }
 darwinia-message-gadget  = { default-features = false, git = "https://github.com/darwinia-network/darwinia-common", branch = "darwinia-v0.12.5" }
 darwinia-staking         = { default-features = false, git = "https://github.com/darwinia-network/darwinia-common", branch = "darwinia-v0.12.5" }
 darwinia-support         = { default-features = false, git = "https://github.com/darwinia-network/darwinia-common", branch = "darwinia-v0.12.5" }
+to-parachain-backing     = { default-features = false, git = "https://github.com/darwinia-network/darwinia-common", branch = "darwinia-v0.12.5" }
```
Comment (HackFisher) [inline runtime/darwinia/Cargo.toml:42]: Can be removed?

File: runtime/darwinia/Cargo.toml [hunk 4]
```diff
@@ -141,6 +148,7 @@ std = [
 	"darwinia-message-gadget/std",
 	"darwinia-staking/std",
 	"darwinia-support/std",
+	"to-parachain-backing/std",
```
Comment (HackFisher) [inline runtime/darwinia/Cargo.toml:151]: To remove

**Comment by HackFisher** [comment]
Pending for migration parameters @xiaoch05

File: runtime/darwinia/src/pallets/bridge_dispatch.rs [hunk 5]
```diff
@@ -126,3 +128,15 @@ impl Config<WithCrabDispatch> for Runtime {
 	type TargetChainAccountPublic = bp_darwinia::AccountPublic;
 	type TargetChainSignature = bp_darwinia::Signature;
 }
+impl Config<WithDarwiniaParachainDispatch> for Runtime {
+	type AccountIdConverter = bp_darwinia_parachain::AccountIdConverter;
```
Comment (AsceticBear) [inline runtime/darwinia/src/pallets/bridge_dispatch.rs:132]: This should be `bp_darwinia::AccountIdConverter` source chain params.

File: runtime/darwinia/src/pallets/bridge_dispatch.rs [hunk 6]
```diff
@@ -126,3 +128,15 @@ impl Config<WithCrabDispatch> for Runtime {
...
+	type BridgeMessageId = (LaneId, MessageNonce);
+	type Call = Call;
+	type CallValidator = CallValidator;
+	type EncodedCall = bm_darwinia_parachain::FromDarwiniaParachainEncodedCall;
+	type Event = Event;
+	type IntoDispatchOrigin = IntoDispatchOrigin;
+	type SourceChainAccountId = bp_darwinia_parachain::AccountId;
+	type TargetChainAccountPublic = bp_darwinia::AccountPublic;
+	type TargetChainSignature = bp_crab::Signature;
```
Comment (AsceticBear) [inline runtime/darwinia/src/pallets/bridge_dispatch.rs:141]: This should be `type TargetChainSignature = bp_darwinia::Signature;`. This dispatch module works on the darwinia side. The target chain is `darwinia`, source chain refers to `darwinia-parachain`.

**Review by AsceticBear — APPROVED** [review]

**Pull request merged**